### DISCUSSION TITLE
After configuring tx-interval/tx-hold using lldpcli, lldpd stops sending/receiving PDUs

### DISCUSSION THREAD

**Discussion by bhinin** [lldpd/lldpd#543] [Q&A] [answered]
I am using lldpcli to configure a custom tx-interval on lldpd 1.0.8. Soon after this configuration, I see lldpd stops sending and receiving lldpd frames. I verified this using `show stat ports <interface name>`

```
[lldpcli] # show stat ports eth0
-------------------------------------------------------------------------------
LLDP statistics:
-------------------------------------------------------------------------------
Interface:    eth0
  Transmitted:  1
  Received:     0
  Discarded:    0
  Unrecognized: 0
  Ageout:       0
  Inserted:     0
  Deleted:      0
-------------------------------------------------------------------------------
```

The one frame sent out is the one advertising TTL = 120 seconds (default tx-interval), no other PDUs are seen. On enabling debugging, I do not see and lldpd logs indicating that the packets are dropped/sent/malformed.

As expected, I also don't see the packets on tcpdump.

I can reliably reproduce this issue, by restarting lldpd, and setting the tx-interval immediately after the lldpd daemon is up. This behavior is seen even when setting tx-hold.

I looked at similar discussions and came across [this](https://github.com/lldpd/lldpd/discussions/485). As suggested, I cherry-picked this [bug-fix to not set VLAN tag](https://github.com/lldpd/lldpd/commit/261afbe371ab316a4bf710338f6d9183a01e083f), but that didn't work.

Also, I am using lldpcli to set other parameters (custom-tlv), but this issue is only observed when setting tx-interval/tx-hold.

Appreciate any help with this!

**Comment by vincentbernat** [comment, marked as answer]
1. Can you reproduce by starting lldpd, wait 5 seconds and set tx-interval?
2. Are you setting tx-interval in lldpd.conf or are you using lldpcli independently?
3. Can you fix the issue by issuing the resume command?

**Reply by bhinin** [reply]
Thanks for the prompt reply.

1. No, the issue is not reproducible after a 5 second wait.
2. In my actual implementation, I am setting the tx-interval using lldpctl and lldpctl_k_config_tx_interval_ms. To isolate and reproduce the issue, I am using lldpcli.
3. This worked. The PDUs were sent out/received after running resume.

**Reply by vincentbernat** [reply]
You run into a race condition where lldpd is running lldpcli to apply initial configuration, then ask to resume. Your program captures the current configuration before the resume command and ask for an update after the resume command. The update contains the "paused" state. You can avoid lldpd to try to autoconfigure with `-L ""`. You can also use the same lock file as lldpcli to avoid concurrent run with lldpcli (the lock is the socket name + `.lock` with a posix lock on it. The file may not exist (created first time lldpcli executes itself), so you may have to create it.

**Reply by bhinin** [reply]
Thank you.

I tried the first suggestion by running lldpd with `-L ""`, but I can still see the issue after setting the tx-interval. I will look more into using the lock used by lldpcli.

Can you elaborate more on the race condition I am hitting? My understanding from what you said :

1. lldpd runs lldpcli to apply initial configuration. The sequence is `"pause" ->" apply config" -> "resume"`
2. My program captures the current configuration in the paused state using `lldpctl_get_configuration()`, sets the tx-interval, then requests an update. Since the current configuration captured above was in paused state, the update requested ends up pausing lldpd again.


Am I correct in my understanding?

**Reply by vincentbernat** [reply]
Yes, this is correct. But since you still got the issue with `-L ""`, I am unsure. The locking discipline would still be useful, but maybe you can just set the unpaused state from your own program to workaround the issue.

**Reply by bhinin** [reply]
Appreciate the quick response times. I am now checking if `lldpctl_k_config_paused` is set, and setting it to false. That worked.
